File: typescript/test/openbao.test.ts
import { createServer, type IncomingMessage, type Server, type ServerResponse } from "node:http";
import { generateKeyPairSync, sign as cryptoSign, type KeyObject } from "node:crypto";
import type { AddressInfo } from "node:net";

import { decodeProtectedHeader, importJWK, jwtVerify } from "jose";
import { afterEach, beforeEach, describe, expect, it } from "vitest";

import { ClientAttestationBuilder, OpenBaoTransitSigner, SigningKeyPair } from "../src/index.js";

const KEY_NAME = "attestation-es256";
const TOKEN = "unit-test-token";

/** In-process fake of the OpenBao transit API surface, holding a real P-256 key. */
class FakeBao {
  readonly token: string;
  readonly #server: Server;
  readonly #privateKey: KeyObject;
  readonly #publicKey: KeyObject;
  #port = 0;

  constructor(token = TOKEN) {
    this.token = token;
    const { privateKey, publicKey } = generateKeyPairSync("ec", { namedCurve: "P-256" });
    this.#privateKey = privateKey;
    this.#publicKey = publicKey;
    this.#server = createServer((req, res) => this.#handle(req, res));
  }

  async start(): Promise<void> {
    await new Promise<void>((resolve) => this.#server.listen(0, "127.0.0.1", resolve));
    this.#port = (this.#server.address() as AddressInfo).port;
  }

  get url(): string {
    return `http://127.0.0.1:${this.#port}`;
  }

  close(): void {
    this.#server.close();
  }

  #handle(req: IncomingMessage, res: ServerResponse): void {
    if (req.headers["x-vault-token"] !== this.token) {
      res.writeHead(403, { "Content-Type": "application/json" });
      res.end('{"errors":["permission denied"]}');
      return;
    }
    if (req.method === "GET") {
      const pem = this.#publicKey.export({ type: "spki", format: "pem" }) as string;
      res.writeHead(200, { "Content-Type": "application/json" });
      res.end(JSON.stringify({ data: { type: "ecdsa-p256", latest_version: 1, keys: { "1": { public_key: pem } } } }));
      return;
    }
    let body = "";
    req.on("data", (chunk: Buffer) => (body += chunk));
    req.on("end", () => {
      const parsed = JSON.parse(body || "{}");
      if (parsed.marshaling_algorithm !== "jws") {
        res.writeHead(400, { "Content-Type": "application/json" });
        res.end('{"errors":["expected marshaling_algorithm=jws"]}');
        return;
      }
      const input = Buffer.from(parsed.input, "base64");
      const raw = cryptoSign("sha256", input, { key: this.#privateKey, dsaEncoding: "ieee-p1363" });
      res.writeHead(200, { "Content-Type": "application/json" });
      res.end(JSON.stringify({ data: { signature: `vault:v1:${raw.toString("base64url")}`, key_version: 1 } }));
    });
  }
}

describe("OpenBao transit signer", () => {
  let bao: FakeBao;

  beforeEach(async () => {
    bao = new FakeBao();
    await bao.start();
  });

  afterEach(() => bao.close());

  it("derives an EC public JWK and kid from key metadata", async () => {
    const signer = await OpenBaoTransitSigner.create(bao.url, TOKEN, KEY_NAME);
    const jwk = signer.publicJwk();
    expect(signer.algorithm).toBe("ES256");
    expect(jwk.kty).toBe("EC");
    expect(jwk.crv).toBe("P-256");
    expect(jwk.kid).toBe(signer.keyId);
    expect(jwk.alg).toBe("ES256");
  });

  it("signs an attestation inside the vault that verifies against the derived key", async () => {
    const signer = await OpenBaoTransitSigner.create(bao.url, TOKEN, KEY_NAME);
    const instance = await SigningKeyPair.generate("ES256");
    const attestation = await new ClientAttestationBuilder(signer, "https://attester.example.com")
      .clientId("https://rp.example.com")
      .confirmationKey(instance)
      .expiresIn(300)
      .build();

    const publicKey = await importJWK(signer.publicJwk(), "ES256");
    const { payload } = await jwtVerify(attestation, publicKey);
    expect(payload.iss).toBe("https://attester.example.com");
    expect(payload.sub).toBe("https://rp.example.com");
    expect((payload.cnf as { jwk: { x: string } }).jwk.x).toBe(instance.publicJwk().x);

    const header = decodeProtectedHeader(attestation);
    expect(header.typ).toBe("oauth-client-attestation+jwt");
    expect(header.kid).toBe(signer.keyId);
  });

  it("fails closed on a wrong token", async () => {
    await expect(OpenBaoTransitSigner.create(bao.url, "wrong-token", KEY_NAME)).rejects.toThrow();
  });

  it("fails closed when the vault is unreachable", async () => {
    const signer = await OpenBaoTransitSigner.create(bao.url, TOKEN, KEY_NAME);
    bao.close();
    await expect(signer.sign(new TextEncoder().encode("header.payload"))).rejects.toThrow();
  });
});

File: go/openbao_test.go
package clientattestation

import (
	"crypto/ecdsa"
	"crypto/elliptic"
	"crypto/rand"
	"crypto/sha256"
	"crypto/x509"
	"encoding/base64"
	"encoding/json"
	"encoding/pem"
	"io"
	"net/http"
	"net/http/httptest"
	"testing"

	"github.com/lestrrat-go/jwx/v2/jwa"
	"github.com/lestrrat-go/jwx/v2/jwk"
	"github.com/lestrrat-go/jwx/v2/jws"
)

const fakeKeyName = "attestation-es256"

// newFakeBao starts an in-process fake of the OpenBao transit API, holding a real P-256 key so produced
// signatures genuinely verify.
func newFakeBao(t *testing.T, token string) (*httptest.Server, *ecdsa.PrivateKey) {
	t.Helper()
	priv, err := ecdsa.GenerateKey(elliptic.P256(), rand.Reader)
	if err != nil {
		t.Fatalf("generate key: %v", err)
	}
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if r.Header.Get("X-Vault-Token") != token {
			w.WriteHeader(http.StatusForbidden)
			io.WriteString(w, `{"errors":["permission denied"]}`)
			return
		}
		if r.Method == http.MethodGet {
			der, _ := x509.MarshalPKIXPublicKey(&priv.PublicKey)
			pemStr := string(pem.EncodeToMemory(&pem.Block{Type: "PUBLIC KEY", Bytes: der}))
			_ = json.NewEncoder(w).Encode(map[string]any{"data": map[string]any{
				"type": "ecdsa-p256", "latest_version": 1,
				"keys": map[string]any{"1": map[string]any{"public_key": pemStr}}}})
			return
		}
		var body map[string]any
		_ = json.NewDecoder(r.Body).Decode(&body)
		if body["marshaling_algorithm"] != "jws" {
			w.WriteHeader(http.StatusBadRequest)
			io.WriteString(w, `{"errors":["expected marshaling_algorithm=jws"]}`)
			return
		}
		input, _ := base64.StdEncoding.DecodeString(body["input"].(string))
		digest := sha256.Sum256(input)
		rInt, sInt, err := ecdsa.Sign(rand.Reader, priv, digest[:])
		if err != nil {
			w.WriteHeader(http.StatusInternalServerError)
			return
		}
		raw := make([]byte, 64)
		rInt.FillBytes(raw[:32])
		sInt.FillBytes(raw[32:])
		envelope := "vault:v1:" + base64.RawURLEncoding.EncodeToString(raw)
		_ = json.NewEncoder(w).Encode(map[string]any{"data": map[string]any{"signature": envelope, "key_version": 1}})
	}))
	return srv, priv
}

func TestOpenBaoDerivesPublicJWK(t *testing.T) {
	srv, _ := newFakeBao(t, "tok")
	defer srv.Close()
	signer, err := NewOpenBaoTransitSigner(srv.URL, "tok", fakeKeyName)
	if err != nil {
		t.Fatalf("new signer: %v", err)
	}
	pub := signer.PublicJWK()
	if signer.Algorithm() != "ES256" {
		t.Errorf("algorithm = %q", signer.Algorithm())
	}
	if pub["kty"] != "EC" || pub["crv"] != "P-256" {
		t.Errorf("jwk = %v", pub)
	}
	if pub["kid"] != signer.KeyID() || pub["alg"] != "ES256" {
		t.Errorf("kid/alg = %v", pub)
	}
}

func TestOpenBaoSignedAttestationVerifies(t *testing.T) {
	srv, _ := newFakeBao(t, "tok")
	defer srv.Close()
	signer, err := NewOpenBaoTransitSigner(srv.URL, "tok", fakeKeyName)
	if err != nil {
		t.Fatalf("new signer: %v", err)
	}
	instance, err := Generate("ES256")
	if err != nil {
		t.Fatalf("generate instance: %v", err)
	}
	attestation, err := NewClientAttestationBuilderWithSigner(signer, "https://attester.example.com").
		ClientID("https://rp.example.com").
		ConfirmationKey(instance).
		ExpiresIn(300).
		Build()
	if err != nil {
		t.Fatalf("build: %v", err)
	}

	pubJSON, _ := json.Marshal(signer.PublicJWK())
	pubKey, err := jwk.ParseKey(pubJSON)
	if err != nil {
		t.Fatalf("parse public jwk: %v", err)
	}
	payload, err := jws.Verify([]byte(attestation), jws.WithKey(jwa.ES256, pubKey))
	if err != nil {
		t.Fatalf("vault-signed attestation did not verify: %v", err)
	}
	var claims map[string]any
	if err := json.Unmarshal(payload, &claims); err != nil {
		t.Fatalf("claims: %v", err)
	}
	if claims["iss"] != "https://attester.example.com" || claims["sub"] != "https://rp.example.com" {
		t.Errorf("claims = %v", claims)
	}
	msg, _ := jws.Parse([]byte(attestation))
	hdr := msg.Signatures()[0].ProtectedHeaders()
	if hdr.Type() != "oauth-client-attestation+jwt" {
		t.Errorf("typ = %q", hdr.Type())
	}
	if hdr.KeyID() != signer.KeyID() {
		t.Errorf("kid = %q", hdr.KeyID())
	}
}

func TestOpenBaoWrongTokenFailsClosed(t *testing.T) {
	srv, _ := newFakeBao(t, "tok")
	defer srv.Close()
	if _, err := NewOpenBaoTransitSigner(srv.URL, "wrong-token", fakeKeyName); err == nil {
		t.Error("a wrong token must fail closed")
	}
}

func TestOpenBaoVaultDownFailsClosed(t *testing.T) {
	srv, _ := newFakeBao(t, "tok")
	signer, err := NewOpenBaoTransitSigner(srv.URL, "tok", fakeKeyName)
	if err != nil {
		t.Fatalf("new signer: %v", err)
	}
	srv.Close()
	if _, err := signer.Sign([]byte("header.payload")); err == nil {
		t.Error("an unreachable vault must fail closed")
	}
}
